geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/utils/ThreadManagement.scala

```diff
@@ -28,28 +36,120 @@ object ThreadManagement extends LazyLogging {
   }
 
   /**
-   * Register a query with the thread manager
+   * Register a scan with the thread manager
+   *
+   * @param scan scan to terminate
+   * @return
    */
-  def register(query: ManagedQuery): ScheduledFuture[_] =
-    executor.schedule(new QueryKiller(query), query.getTimeout, TimeUnit.MILLISECONDS)
+  def register(scan: ManagedScan[_]): ScheduledFuture[_] = {
+    val timeout = math.max(1, scan.timeout.absolute - System.currentTimeMillis())
+    executor.schedule(new QueryKiller(scan), timeout, TimeUnit.MILLISECONDS)
+  }
 
   /**
-    * Trait for classes to be managed for timeouts
-    */
-  trait ManagedQuery extends Closeable {
-    def getTimeout: Long
-    def isClosed: Boolean
-    def debug: String
-  }
+   * Trait for scans that are managed, i.e. tracked and terminated if they exceed a specified timeout
+   *
+   * @tparam T type
+   */
+  trait ManagedScan[T] extends CloseableIterator[T] {
+
+    /**
+     * Scan timeout
+     *
+     * @return
+     */
+    def timeout: Timeout
+
+    /**
+     * Low-level scan to be stopped
+     *
+     * @return
+     */
+    protected def underlying: LowLevelScanner[T]
+
+    // used for log messages
+    protected def typeName: String
+    protected def filter: Option[Filter]
+
+    // we can use a volatile var since we only update the value with a single thread
+    @volatile
+    private var terminated = timeout.absolute <= System.currentTimeMillis()
+
+    private val iter = ExceptionalIterator(if (terminated) { Iterator.empty } else { underlying.iterator })
+    private val cancel = if (terminated) { None } else { Some(ThreadManagement.register(this)) }
```

Comment:
Ok, separate thread/comment...

Is there any daylight between System.currentTimeMillis between these two calls?  Is that a corner case even worth thinking through?  (E.g., could one of them see 'terminated' true and the other not?)

It'd look nutty, but would something like....

val (iter, cancel) = { 
    if (terminated) { (Iterator.empty, None) } 
    else { ... }
} 

make sense?

Reply:
terminated it set and then compared twice, but it can't change between the 2 comparisons b/c it's a var not a def. the call to `ThreadManagement.register(this)` could potentially change the state of terminated, but it's the last thing we do so it shouldn't affect anything in the initialization. and in the register function, we do a `math.max(1, System.currentTimeMillis - timeout)` to ensure that if there is a gap in the sys time we still execute the termination call.

Reply:
i did initially have it in a single if statement, but it seems easier to grok the code this way...

Reply:
I can't read... the var is calculated once.  Mea culpea.  

I'm just searching for any corner case I can make up.  I'll shut up and get back to testing;).